### PR TITLE
Implement a strictly typed service client for comms between services

### PR DESCRIPTION
**Pull request by rikukissa**
This is not used anywhere yet, but the more we start using it the better. This will effectively provide a much nicer dev experience, less room for human errors and better type safety for refactoring.

## Features

**Autocomplete for request paths**
image

**Compiler errors if you are submitting an invalid or missing body**
image

**Typed path parameters**
image

**Typed return values**
image

**Changed the factory call signature a bit**
Service URL now has to be passed from outside
image

#### What's missing
- Typed query parameters
- Quite a lot of Joi definitions from many of our endpoints 😄

### CODE REVIEW

File: packages/commons/src/http.ts [hunk 1]
```diff
@@ -31,3 +33,98 @@ export function joinURL(base: string, path: string) {
...
+  > = (Service[RequestMethod][Path] extends { request: infer R }
+    ? IsNever<R> extends true
+      ? {}
+      : { body: R }
+    : {}) &
+    (Service[RequestMethod][Path] extends { params: infer P }
+      ? IsNever<P> extends true
+        ? {}
+        : { params: P }
+      : {})
```
Comment (naftis) [inline packages/commons/src/http.ts:87]: comment what this does

**Comment by rikukissa** [comment]
@naftis @Zangetsu101  I've also started thinking if writing our internal APIs in https://trpc.io/ would actually be better in the grand scheme of things. Thoughts?

**Comment by Zangetsu101** [comment]
> @naftis @Zangetsu101 I've also started thinking if writing our internal APIs in https://trpc.io/ would actually be better in the grand scheme of things. Thoughts? It would take us a step further from "easily approachable tech" but would yield other benefits for our team

tRPC is ideal for communicating between frontend and backend so kinda at the place where our GraphQL layer sits. Of course it has this requirement that both of them must be using typescript (which we already do). But not sure if anyone's using it for backend services tho.  TBH this PR did look to me like a tRPC implementation but for backend microservices :smile:

**Comment by naftis** [comment]
Idea: the country-config routes could be typed like this as well. Then we could go as far as create a Swagger / OpenAPI docs of what routes the country-config should implement.

**Comment by rikukissa** [comment]
I'm closing this for now. I think tRPC is a better option

**Pull request closed without merging**